src/tests/unit_x509.cpp

```diff
@@ -1027,6 +1028,27 @@ Test::Result test_usage(const Botan::Private_Key& ca_key,
    result.confirm("key usage digitalSignature allowed", no_usage_cert.allowed_usage(Key_Constraints::DIGITAL_SIGNATURE));
    result.confirm("key usage cRLSign allowed", no_usage_cert.allowed_usage(Key_Constraints::CRL_SIGN));
 
+   if (sig_algo == "RSA")
```

Comment:
I'd like to see some additional tests here, using the original constraints and for all key types:

- If key encipherment or data encipherment are set, test that ENCRYPTION is allowed
- Otherwise, test that ENCRYPTION is denied

As it is (with the current tests) allowed_usage(ENCRYPTION) could simply be `return true` regardless of usage or algorithm and everything would pass.

[The test you have already written is also fine to include.]

Reply:
I added 'negative' tests for the already existing constraints. None of those previously neither allowed  `KEY_ENCIPHERMENT` nor `DATA_ENCIPHERMENT`. The "cert allows every usage" [above](https://github.com/randombit/botan/pull/2106/files#r327026610) is a bit questionable though.